torch/nn/utils/stateless.py

```diff
@@ -37,144 +45,169 @@ def _setattr(self, name: str, value: Any) -> None:
-    return _swap_parameters
+    Returns:
+        A new untied version of the parameters_and_buffers dictionary.
 
+    Raises:
+        ValueError: if there are more than one user-given values for the same tied tensor.
+    """
+    # A map of {name: tensor} for all tensors (including tied ones) in the module.
+    all_named_tensors: Dict[str, Tensor] = {}
+    all_named_tensors.update(module.named_parameters(remove_duplicate=False))
+    all_named_tensors.update(module.named_buffers(remove_duplicate=False))
+
+    # A map of {tensor: set(all_tied_names)} for all tensor names in the module.
+    tensor_to_tied_names_map: Dict[Tensor, Set[str]] = defaultdict(set)
+    for name, tensor in all_named_tensors.items():
+        tensor_to_tied_names_map[tensor].add(name)
+
+    # A map of {tied_name: set(all_tied_names)} for all tensor names in the module.
+    # If a name is not tied, it will not be in this map.
+    tied_names_map: Dict[str, Set[str]] = {}
+    for tied_names in tensor_to_tied_names_map.values():
+        if len(tied_names) > 1:
+            for tied_name in tied_names:
+                tied_names_map[tied_name] = tied_names
+
+    # Make sure the user didn't pass multiple values for the same tied tensor.
+    given_names = set(parameters_and_buffers.keys())
+    given_names_for_tied_tensors = given_names.intersection(tied_names_map.keys())
+    for given_name in given_names_for_tied_tensors:
+        tied_names = tied_names_map[given_name]
+        if (
+            # Detect if there are multiple keys present for the same tied tensor.
+            len(tied_names.intersection(given_names_for_tied_tensors)) > 1
+            # Only raise an error if the user passed multiple values for the same tied tensor.
+            # If all given values are the same, don't raise.
+            and len({parameters_and_buffers[tied_name] for tied_name in tied_names})
+            != 1
+        ):
+            raise ValueError(
+                f"functional_call got multiple values for keys {sorted(tied_names)}, "
+                f"which are tied. Consider using tie_weights=False"
+            )
 
-def _remove_swap(module, name: str, full_path: str) -> None:
-    if hasattr(module, "_orig_class"):
-        module.__class__ = module._orig_class
-        delattr(module, "_orig_class")
-        delattr(module, "_attr_to_path")
+    # Untie the given named tensor map
+    # Make a copy for not modifying the original dict
+    untied_parameters_and_buffers = parameters_and_buffers.copy()
+    for given_name in given_names_for_tied_tensors:
+        for tied_name in tied_names_map[given_name]:
+            untied_parameters_and_buffers[tied_name] = parameters_and_buffers[
+                given_name
+            ]
+    return untied_parameters_and_buffers
 
 
 @contextlib.contextmanager
 def _reparametrize_module(
-    module: 'torch.nn.Module',
+    module: "torch.nn.Module",
     parameters_and_buffers: Dict[str, Tensor],
+    *,
```

Comment:
I think making `tie_weights` a kw-only arg is technically BC-breaking, since it affects the signature of this function which is imported as public into `torch.nn.utils._stateless` (see [here](https://github.com/pytorch/pytorch/blob/e52786f3d177a7ca5d490a516cf52e236ef072cb/torch/nn/utils/_stateless.py#L15)).

That said, I don't think we care too much since `_stateless` is deprecated. cc @albanD wdyt? when do we plan to stop caring about `_stateless` BC?

Reply:
I think this change is minor because `tie_weights` defaults to be `True`. If users do not want to tie weights, they can simply pass a full `parameters_and_buffers` dict rather than set `tie_weights=False`.

Reply:
It has been long enough IIRC and we should consider removing `_stateless` altogether.

Reply:
I removed `torch.nn.utils._stateless` in the newest commit. It has been deprecated for 10 months.

Reply:
@XuehaiPan, do you mind bumping the removal of `torch.nn.utils._stateless` to a follow-up PR? I think we want to keep the scope of this PR to the refactor. Also, we generally like to make hard-breaking changes in isolation so they're easier to revert if needed.

Reply:
I reverted the breaking change of the file removal and kwonly `tie_weights`.